### PR TITLE
[FLINK-35109] Drop support for Flink 1.17 & 1.18 and fix tests for 1.20-SNAPSHOT

### PR DESCRIPTION
**Pull request by ruanhang1993**
This PR drops support for Flink 1.17 & 1.18 and fix tests for 1.20-SNAPSHOT.

### CODE REVIEW

**Comment by ruanhang1993** [comment]
@MartijnVisser @fapaul Please help to review this PR, thanks~

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
I think there are some things we need to account for still

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/TypeSerializerUpgradeTestBase.java [hunk 1]
```diff
@@ -43,28 +40,19 @@
...
- */
+/** A test base for testing {@link TypeSerializer} upgrades. */
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
-public abstract class TypeSerializerUpgradeTestBase<PreviousElementT, UpgradedElementT> {
-
-    public static final FlinkVersion CURRENT_VERSION = FlinkVersion.v1_17;
-
-    public static final Set<FlinkVersion> MIGRATION_VERSIONS =
-            FlinkVersion.rangeOf(FlinkVersion.v1_11, CURRENT_VERSION);
+public abstract class TypeSerializerUpgradeTestBase<PreviousElementT, UpgradedElementT>
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/TypeSerializerUpgradeTestBase.java:54]: Hi, @MartijnVisser .
The changes for `TypeSerializerUpgradeTestBase` in #24603 and #23960 make the kafka connector not be able to compile with both 1.20-SNAPSHOT and 1.19.0.

I think we still have to maintain `TypeSerializerUpgradeTestBase`.

File: .github/workflows/push_pr.yml [hunk 2]
```diff
@@ -28,21 +28,16 @@ jobs:
...
       matrix:
-        flink: [ 1.17.2 ]
-        jdk: [ '8, 11' ]
-        include:
-          - flink: 1.18.1
-            jdk: '8, 11, 17'
-          - flink: 1.19.0
-            jdk: '8, 11, 17, 21'
+        flink: [ 1.19.0, 1.20-SNAPSHOT ]
+        jdk: [ '8, 11, 17, 21' ]
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply .github/workflows/push_pr.yml:32]: Hi, @MartijnVisser .
I check these files in both jdbc connector and mongodb connector. It seems like they all use the Java 21 for Flink 1.19. Do I need to change them too?

[the remark this answers is not in the document]
Comment (snuyanzin) [reply .github/workflows/push_pr.yml:32]: May be I didn't get your message about java 21 support, however it is mentioned in 1.19 release notes [1]
Also we have nightly with jdk21 tests (same as for jdk17) starting 1.19 e.g. [2]

[1] https://flink.apache.org/2024/03/18/announcing-the-release-of-apache-flink-1.19/#beta-support-for-java-21
[2] https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=59560&view=results

[the remark this answers is not in the document]
Comment (MartijnVisser) [reply .github/workflows/push_pr.yml:32]: > May be I didn't get your message about java 21 support, however it is mentioned in 1.19 release notes

🤦 I mixed up Flink 1.18 and 1.19, sorry. My comment was incorrect

**Pull request closed without merging**